src/sources/opentelemetry/http.rs

```diff
@@ -324,56 +359,74 @@ async fn handle_request(
-                warp::reject::custom(ApiError::ServerShutdown)
-            })?;
-
-            match receiver {
-                None => Ok(protobuf(resp).into_response()),
-                Some(receiver) => match receiver.await {
-                    BatchStatus::Delivered => Ok(protobuf(resp).into_response()),
-                    BatchStatus::Errored => Err(warp::reject::custom(Status {
-                        code: 2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
-                        message: "Error delivering contents to sink".into(),
-                        ..Default::default()
-                    })),
-                    BatchStatus::Rejected => Err(warp::reject::custom(Status {
-                        code: 2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
-                        message: "Contents failed to deliver to sink".into(),
-                        ..Default::default()
-                    })),
-                },
-            }
-        }
-        Err(err) => Err(warp::reject::custom(err)),
+    let reply_with_status =
+        |http_code: hyper::StatusCode, code: i32, message: String| -> Result<Response, Rejection> {
+            let mut resp = protobuf(Status {
+                code,
+                message,
+                ..Default::default()
+            })
+            .into_response();
+            *resp.status_mut() = http_code;
+            Ok(resp)
+        };
+
+    if let Err(err) = events {
+        return reply_with_status(
+            err.status_code(),
+            2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
+            err.message().into(),
+        );
     }
-}
 
-async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::Infallible> {
-    if let Some(err_msg) = err.find::<ErrorMessage>() {
-        let reply = protobuf(Status {
-            code: 2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
-            message: err_msg.message().into(),
-            ..Default::default()
-        });
+    let mut events = events.unwrap();
+    let receiver = BatchNotifier::maybe_apply_to(acknowledgements, &mut events);
+    let count = events.len();
+
+    if let Err(_) = out.send_batch_named(output, events).await {
+        emit!(StreamClosedError { count });
+        // the client can try again later
+        return reply_with_status(
+            hyper::StatusCode::SERVICE_UNAVAILABLE,
+            2,
+            "Server is shutting down".into(),
+        );
+    };
+
+    match receiver {
+        None => Ok(protobuf(resp).into_response()),
+        Some(receiver) => match receiver.await {
+            BatchStatus::Delivered => Ok(protobuf(resp).into_response()),
+            BatchStatus::Errored => reply_with_status(
+                hyper::StatusCode::INTERNAL_SERVER_ERROR,
```

Comment:
Is there a way to tell why it errored and whether it makes sense to return a 'retriable' status code (e.g. 502)